### PR TITLE
Shrsha/fix bug ts collect

### PR DESCRIPTION
**Pull request by shreeyaharma**
Test job https://ml.azure.com/experiments/id/4b4e34ef-331f-4b1a-b1b0-88615d6c45ad/runs/lemon_deer_kqsrzgpg25?wsid=/subscriptions/1aefdc5e-3a7c-4d71-a9f9-f5d3b03be19a/resourcegroups/TestingTeamProjectRG/providers/Microsoft.MachineLearningServices/workspaces/shrsha-eastus-test&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

Output from main branch to prove there is no diff in output after logic change:
metrics calculated
+--------------------+------------------+---------------+
|         metric_name|      metric_value|threshold_value|
+--------------------+------------------+---------------+
|     TotalTokenCount|            4538.0|            0.0|
|    TotalPromptCount|            4441.0|            0.0|
|TotalCompletionCount|              97.0|            0.0|
|       AvgTokenCount| 648.2857142857143|            0.0|
|      AvgPromptCount| 634.4285714285714|            0.0|
|  AvgCompletionCount|13.857142857142858|            0.0|
|gpt-35-turbo-inst...|              97.0|            0.0|
|gpt-35-turbo-inst...|               4.0|            0.0|
|gpt-35-turbo-inst...|            4424.0|            0.0|
| ada.CompletionCount|               0.0|            0.0|
|   ada.TotalRequests|               3.0|            0.0|
|     ada.PromptCount|              17.0|            0.0|
+--------------------+------------------+---------------+

Using AML OBO credential from StoreUrl.get_credential() with blob datastore whose credential type is null.
_AzureMLSparkOnBehalfOfCredential.get_token succeeded
_AzureMLSparkOnBehalfOfCredential.get_token succeeded
+-----------+--------------------+---------------+--------------------+
|PromptCount|               Input|CompletionCount|              Output|
+-----------+--------------------+---------------+--------------------+
|       69.0|{\n  "prompt": "s...|            7.0|" Can you repeat ...|
|     1674.0|{\n  "prompt": "{...|           25.0|" I'm sorry, I do...|
|       72.0|{\n  "prompt": "s...|            9.0|" Can you explain...|
|        1.0|{\n  "mlindex_con...|            0.0|[\n  {\n    "text...|
|     2609.0|{\n  "prompt": "{...|           56.0|" A feature store...|
+-----------+--------------------+---------------+--------------------+

### CODE REVIEW

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py [hunk 1]
```diff
@@ -54,127 +38,103 @@ def create_span_tree_from_dataframe_row(self, row: Row) -> Optional[SpanTreeNode
     def create_span_tree_from_dataframe(self, df: DataFrame) -> List[Optional[SpanTreeNode]]:
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:38]: Is this function still resued?

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py [hunk 2]
```diff
@@ -54,127 +38,103 @@ def create_span_tree_from_dataframe_row(self, row: Row) -> Optional[SpanTreeNode
...
         """Create a list of SpanTrees from a DataFrame."""
         # Convert the DataFrame to an RDD and then to a list of Rows
-        rows = df.rdd.map(lambda row: row).collect()
-
-        # Create a SpanTree for each row
-        span_trees = [self.create_span_tree_from_dataframe_row(row) for row in rows]
-        return span_trees
+        df.rdd.foreach(self.process_row)
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:41]: I don't see anything returned, according to the signature, there should be something returned

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py [hunk 3]
```diff
@@ -54,127 +38,103 @@ def create_span_tree_from_dataframe_row(self, row: Row) -> Optional[SpanTreeNode
...
 
     def process(self, df_traces: DataFrame):
         """Process the aggregated trace data to compute metrics.
 
         Args:
             df_traces (DataFrame): Aggregated traces from the GenAI processor.
         """
+        print("process")
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:49]: Can we have more clear, user friendly information/message?

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py [hunk 4]
```diff
@@ -54,127 +38,103 @@ def create_span_tree_from_dataframe_row(self, row: Row) -> Optional[SpanTreeNode
...
-        if root_input is not None and root_output is not None:
-
-            self.request_data.append({
-                "prompt_count": prompt_count,
-                "input": root_input,
-                "completion_count": completion_count,
-                "output": root_output
-            })
+        tree = self.create_span_tree_from_dataframe_row(row)
+        print("tree before", tree)
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:101]: More clear message if we want to have the debug information

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py [hunk 5]
```diff
@@ -54,127 +38,103 @@ def create_span_tree_from_dataframe_row(self, row: Row) -> Optional[SpanTreeNode
...
-    def metrics_generator(self):
-        """Return: metrics_data_df (DataFrame): metrics dataframe."""
-        spark = init_spark()
-        schema = StructType([
-            StructField(SIGNAL_METRICS_METRIC_NAME, StringType(), True),
-            StructField(SIGNAL_METRICS_METRIC_VALUE, DoubleType(), True),
-            StructField(SIGNAL_METRICS_THRESHOLD_VALUE, StringType(), True)
+        # Create an RDD of tuples with the necessary counts
+        counts_and_requests_rdd = df_traces.rdd.flatMap(self.process_row)
+        counts_rdd = counts_and_requests_rdd.filter(lambda x: isinstance(x, tuple))
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:53]: Seems a very fragile contract

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py [hunk 6]
```diff
@@ -54,127 +38,103 @@ def create_span_tree_from_dataframe_row(self, row: Row) -> Optional[SpanTreeNode
...
+                    parent_input = parent.input
+                    parent_output = parent.output
+                # Extract the counts and model information
+                completion_count = self.get_value_from_attributes(attributes, COMPLETION_COUNT_KEYS)
+                prompt_count = self.get_value_from_attributes(attributes, PROMPT_COUNT_KEYS)
+                total_count = self.get_value_from_attributes(attributes, TOTAL_COUNT_KEYS)
+                model = self.get_value_from_attributes(attributes, MODEL_KEYS)
+
+                # Append a tuple with the counts and model to the list
+                if model is not None:
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:124]: Is there any case we don't have model information, and we may still need to add the total token account

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/run.py [hunk 7]
```diff
@@ -29,9 +42,48 @@ def run():
 
     processor = MetricsProcessor()
     processor.process(token_df)
-    metrics_df = processor.metrics_generator()
-    save_spark_df_as_mltable(metrics_df, args.signal_metrics)
-    sample_data = processor.get_latest_run()
+    metrics_df = processor.metrics_df
+
+    flat_metrics_df = metrics_df.select(
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/run.py:47]: What is this logic for?

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/run.py [hunk 8]
```diff
@@ -29,9 +42,48 @@ def run():
...
+                                                            F.col('metrics')[SIGNAL_METRICS_THRESHOLD_VALUE]
+                                                            .alias(SIGNAL_METRICS_THRESHOLD_VALUE)
+                                                            )
+    flat_metrics_df.show()
+    flat_metrics_rdd = processor.metrics_model_df.rdd.flatMap(lambda row: [
+        (f"{row['model']}.CompletionCount", row['total_completion_count'], 0),
+        (f"{row['model']}.TotalRequests", row['total_requests'], 0),
+        (f"{row['model']}.PromptCount", row['total_prompt_count'], 0)
+    ])
+    flat_model_metrics_df = flat_metrics_rdd.toDF([SIGNAL_METRICS_METRIC_NAME, SIGNAL_METRICS_METRIC_VALUE,
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/run.py:81]: Could you please add unit tests for the function?

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py (hunk 3 above)
Comment (shreeyaharma) [reply assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:49]: I'll remove it

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py (hunk 5 above)
Comment (shreeyaharma) [reply assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:53]: they are independent metrics

File: assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py (hunk 6 above)
Comment (shreeyaharma) [reply assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:124]: information for aggregation requires model

Comment (yunjie-hub) [reply assets/model_monitoring/components/src/genai_token_statistics_compute_metrics/metrics_processor.py:118]: For model level metrics we need that, while for total I don't think it's required

**Review by yunjie-hub — APPROVED** [review]

**Pull request merged**